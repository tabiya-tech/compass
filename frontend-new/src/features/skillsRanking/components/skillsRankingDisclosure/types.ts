export const jobSeekerComparisonLabels = ["LOWEST", "SECOND LOWEST", "MIDDLE", "SECOND HIGHEST", "HIGHEST"]