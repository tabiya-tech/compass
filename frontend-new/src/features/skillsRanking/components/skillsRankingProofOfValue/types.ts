export enum EffortType {
  TIME_BASED = "time_based",
  WORK_BASED = "work_based",
}