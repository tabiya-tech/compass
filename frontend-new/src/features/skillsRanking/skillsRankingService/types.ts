export type SkillsRankingMetrics = {
  cancelled_after?: string,
  succeeded_after?:string,
  puzzles_solved?:number,
  correct_rotations?:number,
  clicks_count?:number,
}