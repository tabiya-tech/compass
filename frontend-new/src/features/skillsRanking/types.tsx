export enum CompareAgainstGroup {
  AGAINST_OTHER_JOB_SEEKERS = "against_other_job_seekers",
  AGAINST_JOB_MARKET = "against_job_market"
}

export enum ButtonOrderGroup {
  SKIP_BUTTON_FIRST = "skip_button_first",
  VIEW_BUTTON_FIRST = "view_button_first"
}

export interface SkillRankingExperimentGroups {
  compare_against: CompareAgainstGroup;
  button_order: ButtonOrderGroup;
  delayed_results: boolean;
}

export enum SkillsRankingPhase {
  INITIAL = "INITIAL",
  SKIPPED = "SKIPPED",
  CANCELLED = "CANCELLED",
  SELF_EVALUATING = "SELF_EVALUATING",
  EVALUATED = "EVALUATED"
}

export interface SkillsRankingState {
  session_id: number;
  experiment_groups: SkillRankingExperimentGroups;
  phase: SkillsRankingPhase;
  ranking: string | null;
  self_ranking: string | null;
}

export const skillsRankingStateDefault: SkillsRankingState = {
  session_id: -1,
  experiment_groups: {
    compare_against: CompareAgainstGroup.AGAINST_OTHER_JOB_SEEKERS,
    button_order: ButtonOrderGroup.SKIP_BUTTON_FIRST,
    delayed_results: false,
  },
  phase: SkillsRankingPhase.INITIAL,
  ranking: null,
  self_ranking: null,
}

export interface SkillsRankingResult {
  ranking: string;
}