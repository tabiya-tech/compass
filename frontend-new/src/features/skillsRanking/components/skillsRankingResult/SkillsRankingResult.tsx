import React, { useEffect, useState } from "react";
import { 
  CompareAgainstGroup,
  SkillsRankingState 
} from "src/features/skillsRanking/types";
import { Box, CircularProgress, Typography, useTheme } from "@mui/material";
import { MessageContainer } from "src/chat/chatMessage/compassChatMessage/CompassChatMessage";
import ChatBubble from "src/chat/chatMessage/components/chatBubble/ChatBubble";
import { ConversationMessageSender } from "src/chat/ChatService/ChatService.types";

const uniqueId = "3fb55364-3b63-48db-b1ef-818a4daa5862";

export const SKILLS_RANKING_RESULT_MESSAGE_TYPE = `skills-ranking-result-message-${uniqueId}`;

export const DATA_TEST_ID = {
  SKILLS_RANKING_RESULT_CONTAINER: `skills-ranking-result-container-${uniqueId}`,
  SKILLS_RANKING_RESULT_TEXT: `skills-ranking-result-text-${uniqueId}`,
  SKILLS_RANKING_LOADING_TEXT: `skills-ranking-loading-text-${uniqueId}`,
  SKILLS_RANKING_LOADING_INDICATOR: `skills-ranking-loading-indicator-${uniqueId}`,
  SKILLS_RANKING_ERROR: `skills-ranking-error-${uniqueId}`,
};

export interface SkillsRankingResultProps {
  message: string;
  skillsRankingState: SkillsRankingState;
  isLoading: boolean;
}

export const IMMEDIATE_RESULT_MESSAGES = {
  [CompareAgainstGroup.AGAINST_OTHER_JOB_SEEKERS]: (rank: string) => (
    <>
      Based on how important your skills are versus other job seekers, you currently rank in the <b>{rank}</b> percentile. Do note that
      this can always change as you continue learning!
    </>
  ),
  [CompareAgainstGroup.AGAINST_JOB_MARKET]: (rank: string) => (
    <>
      Based on how important your skills are versus the job market, you currently rank in the <b>{rank}</b> percentile. Do note that this
      can always change as you continue learning!
    </>
  ),
};

export const DELAYED_RESULT_MESSAGE = "Your skills ranking is taking longer than expected to calculate. We will share the results with you via SMS within 24 hours."

const SkillsRankingResult: React.FC<Readonly<SkillsRankingResultProps>> = ({ 
  message, 
  skillsRankingState,
  isLoading
}) => {
  const theme = useTheme();
  const [compareAgainstGroup, setCompareAgainstGroup] = useState<CompareAgainstGroup | null>(null);
  const [delayedResults, setDelayedResults] = useState<boolean>(false);

  useEffect(() => {
    setCompareAgainstGroup(skillsRankingState.experiment_groups.compare_against);
    setDelayedResults(skillsRankingState.experiment_groups.delayed_results);
  }, [skillsRankingState.experiment_groups]);

  console.log('compareAgainstGroup:', compareAgainstGroup);
  console.log('delayedResults:', delayedResults);
  console.log('skillsRankingState:', skillsRankingState);

  return (
    <MessageContainer origin={ConversationMessageSender.COMPASS} data-testid={DATA_TEST_ID.SKILLS_RANKING_RESULT_CONTAINER}>
      <ChatBubble message={delayedResults ? DELAYED_RESULT_MESSAGE : message} sender={ConversationMessageSender.COMPASS}>
        <Box>
          {isLoading ? (
            <Box display="flex">
              <Typography
                variant="body1"
                alignItems="center"
                gap={theme.fixedSpacing(theme.tabiyaSpacing.xs)}
                data-testid={DATA_TEST_ID.SKILLS_RANKING_LOADING_TEXT}
              >
                Please wait while I compare your skillset...
              </Typography>
              <CircularProgress
                size={16}
                aria-label="Analyzing skillset"
                sx={{ color: (theme) => theme.palette.info.contrastText }}
                data-testid={DATA_TEST_ID.SKILLS_RANKING_LOADING_INDICATOR}
              />
            </Box>
          ) : (
            <Typography variant="body1" data-testid={DATA_TEST_ID.SKILLS_RANKING_RESULT_TEXT}>
              {(skillsRankingState.ranking && compareAgainstGroup && !delayedResults) ? IMMEDIATE_RESULT_MESSAGES[compareAgainstGroup](skillsRankingState.ranking) : ""}
            </Typography>
          )}
        </Box>
      </ChatBubble>
    </MessageContainer>
  );
};

export default SkillsRankingResult;
