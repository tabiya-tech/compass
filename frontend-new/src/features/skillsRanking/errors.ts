
export class SkillsRankingError extends Error {
    constructor(message: string, cause?: unknown) {
      super(message);
      this.name = "SkillsRankingError";
      this.cause = cause;
    }
  }