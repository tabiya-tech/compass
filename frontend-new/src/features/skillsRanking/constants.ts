import { SkillsRankingService } from "./skillsRankingService/skillsRankingService";

export const SKILLS_RANKING_FEATURE_ID = "4b0c7428-9c01-4688-81fd-d3ef159bce79";

// Typing durations for different components
export const TYPING_DURATION_MS = 5000;
export const MESSAGE_DURATION_MS = 5000;
export const CALCULATION_DELAY = 60000;
export const EFFORT_METRICS_UPDATE_INTERVAL = 2000;

// Puzzle constants
export const DEFAULT_STRINGS = ["GJRLK", "FQZNC", "EKJGR", "CJFLQ", "GRKLE"];

// Config constants - these values are unlikely to change during runtime
export const getJobPlatformUrl = () => SkillsRankingService.getInstance().getConfig().config.jobPlatformUrl;
export const getAirtimeBudget = () => SkillsRankingService.getInstance().getConfig().config.airtimeBudget;
