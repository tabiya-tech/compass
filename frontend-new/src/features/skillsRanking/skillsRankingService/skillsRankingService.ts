import { getBackendUrl } from "src/envService";
import { customFetch } from "src/utils/customFetch/customFetch";
import { StatusCodes } from "http-status-codes";
import {
  isValidExperimentGroupKey,
  SkillsRankingConfig,
  SkillsRankingExperimentGroups,
  SkillsRankingPhase,
  SkillsRankingState,
} from "src/features/skillsRanking/types";
import { FeaturesService } from "src/features/featuresService/FeaturesService";
import { SKILLS_RANKING_FEATURE_ID } from "src/features/skillsRanking/constants";
import { SkillsRankingError } from "../errors";
import { SkillsRankingMetrics } from "../types";

// Throttled metrics updater for periodic updates during active work
const createThrottledMetricsUpdater = (
  sessionId: number,
  updateFunction: (sessionId: number, metrics: SkillsRankingMetrics) => Promise<any>,
  throttleInterval: number = 5000,
  debounceDelay: number = 2000
) => {
  let lastMetrics: SkillsRankingMetrics | null = null;
  let lastUpdateTime = 0;
  let throttleTimeoutId: ReturnType<typeof setTimeout> | null = null;
  let debounceTimeoutId: ReturnType<typeof setTimeout> | null = null;
  let isActive = false;

  const sendUpdate = async (metrics: SkillsRankingMetrics) => {
    // Don't send if metrics haven't changed
    if (
      lastMetrics &&
      lastMetrics.puzzles_solved === metrics.puzzles_solved &&
      lastMetrics.correct_rotations === metrics.correct_rotations &&
      lastMetrics.clicks_count === metrics.clicks_count
    ) {
      return;
    }

    try {
      await updateFunction(sessionId, metrics);
      lastMetrics = { ...metrics };
      lastUpdateTime = Date.now();
    } catch (err) {
      console.error("Failed to update metrics periodically", err);
    }
  };

  return {
    // Call this when user is actively working
    onActivity: (metrics: SkillsRankingMetrics) => {
      isActive = true;

      // Clear any pending debounce
      if (debounceTimeoutId) {
        clearTimeout(debounceTimeoutId);
        debounceTimeoutId = null;
      }

      const now = Date.now();

      // If enough time has passed since last update, send immediately
      if (now - lastUpdateTime >= throttleInterval) {
        sendUpdate(metrics);
      } else {
        // Schedule a throttled update
        if (throttleTimeoutId) {
          clearTimeout(throttleTimeoutId);
        }
        throttleTimeoutId = setTimeout(
          () => {
            if (isActive) {
              sendUpdate(metrics);
            }
          },
          throttleInterval - (now - lastUpdateTime)
        );
      }
    },

    // Call this when user stops working (for debouncing)
    onInactivity: (metrics: SkillsRankingMetrics) => {
      isActive = false;

      // Clear any pending throttle
      if (throttleTimeoutId) {
        clearTimeout(throttleTimeoutId);
        throttleTimeoutId = null;
      }

      // Debounce the final update
      if (debounceTimeoutId) {
        clearTimeout(debounceTimeoutId);
      }
      debounceTimeoutId = setTimeout(() => {
        sendUpdate(metrics);
      }, debounceDelay);
    },

    // Call this to force an immediate update (e.g., for cancellation)
    forceUpdate: (metrics: SkillsRankingMetrics) => {
      // Clear any pending timeouts
      if (throttleTimeoutId) {
        clearTimeout(throttleTimeoutId);
        throttleTimeoutId = null;
      }
      if (debounceTimeoutId) {
        clearTimeout(debounceTimeoutId);
        debounceTimeoutId = null;
      }

      sendUpdate(metrics);
    },

    // Clean up timeouts
    cleanup: () => {
      if (throttleTimeoutId) {
        clearTimeout(throttleTimeoutId);
        throttleTimeoutId = null;
      }
      if (debounceTimeoutId) {
        clearTimeout(debounceTimeoutId);
        debounceTimeoutId = null;
      }
    },
  };
};

/**
 * Service to manage the skills ranking feature.
 */
export class SkillsRankingService extends FeaturesService {
  private static instance: SkillsRankingService;
  readonly skillsRankingEndpointUrl: string;
  readonly apiServerUrl: string;
  readonly featureId = SKILLS_RANKING_FEATURE_ID;

  private constructor() {
    super(SKILLS_RANKING_FEATURE_ID);
    this.apiServerUrl = getBackendUrl();
    this.skillsRankingEndpointUrl = `${this.apiServerUrl}/${super.getFeatureAPIPrefix(this.featureId)}/conversations`;
  }

  /**
   * Get the singleton instance of the SkillsRankingService.
   * @returns {SkillsRankingService} The singleton instance of the SkillsRankingService.
   */
  static getInstance(): SkillsRankingService {
    if (!SkillsRankingService.instance) {
      SkillsRankingService.instance = new SkillsRankingService();
    }
    return SkillsRankingService.instance;
  }

  /**
   * Check if the skills ranking feature is enabled
   */
  isSkillsRankingFeatureEnabled(): boolean {
    return super.isFeatureEnabled();
  }

  /**
   * Get the ranking state for a given session.
   *
   * @param sessionId - The ID of the session to get the ranking state for.
   * @returns {Promise<SkillsRankingState | null>} The ranking state for the session, or null if none exists.
   */
  async getSkillsRankingState(sessionId: number): Promise<SkillsRankingState | null> {
    const url = `${this.skillsRankingEndpointUrl}/${sessionId}/skills-ranking/state`;

    const response = await customFetch(url, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
      expectedStatusCode: StatusCodes.OK,
      serviceName: "SkillsRankingService",
      serviceFunction: "getSkillsRankingState",
      failureMessage: `Failed to get skills ranking state for session ${sessionId}`,
      expectedContentType: "application/json",
    });

    const data = await response.json();
    if (data === null) {
      return null;
    }

    // Validate experiment group returned from the API.
    if (data.experiment_group && !isValidExperimentGroupKey(data.experiment_group)) {
      throw new SkillsRankingError(`Unknown experiment_group '${data.experiment_group}' from API`);
    }

    // REVIEW: Have a private function to convert the data to SkillsRankingState
    return {
      session_id: data.session_id,
      experiment_group:
        SkillsRankingExperimentGroups[data.experiment_group as keyof typeof SkillsRankingExperimentGroups],
      phases: data.phase,
      score: data.score,
      cancelled_after: data.cancelled_after,
      succeeded_after: data.succeeded_after,
      puzzles_solved: data.puzzles_solved,
      correct_rotations: data.correct_rotations,
      clicks_count: data.clicks_count,
      perceived_rank_percentile: data.perceived_rank_percentile,
      retyped_rank_percentile: data.retyped_rank_percentile,
      started_at: data.started_at,
      completed_at: data.completed_at,
    };
  }

  /**
   * Update the ranking state for a given session.
   *
   * @param sessionId - The ID of the session to update the ranking state for.
   * @param phase - The new current phase.
   *   This is only relevant for the effort-based proof_of_value task.
   *   This is only relevant for the time-based proof_of_value task.
   *     This is only relevant for the time-based proof_of_value task.
   * @param perceived_rank_percentile - Optional parameter indicating the user's perceived rank percentile (0-100).
   * @param retyped_rank_percentile - Optional parameter indicating the rank the user retyped to confirm they saw it correctly (0-100).
   * @param metrics
   * @returns {Promise<SkillsRankingState>} The updated ranking state for the session.
   */
  async updateSkillsRankingState(
    sessionId: number,
    phase: SkillsRankingPhase,
    perceived_rank_percentile?: number,
    retyped_rank_percentile?: number,
    metrics?: SkillsRankingMetrics
  ): Promise<SkillsRankingState> {
    const url = `${this.skillsRankingEndpointUrl}/${sessionId}/skills-ranking/state`;

    // Build request body based on phase validation rules
    const body: any = {
      phase: phase,
    };

    // Add metrics fields only for phases that allow them
    if (phase === SkillsRankingPhase.PROOF_OF_VALUE) {
      body.cancelled_after = metrics?.cancelled_after ?? null;
      body.succeeded_after = metrics?.succeeded_after ?? null;
      body.puzzles_solved = metrics?.puzzles_solved ?? null;
      body.correct_rotations = metrics?.correct_rotations ?? null;
      body.clicks_count = metrics?.clicks_count ?? null;
    }

    // Add rank fields for phases that allow them
    if (phase === SkillsRankingPhase.PERCEIVED_RANK) {
      body.perceived_rank_percentile = perceived_rank_percentile ?? null;
    }

    if (phase === SkillsRankingPhase.RETYPED_RANK) {
      body.retyped_rank_percentile = retyped_rank_percentile ?? null;
    }

    // Allow both rank fields when transitioning to COMPLETED
    if (phase === SkillsRankingPhase.COMPLETED) {
      body.perceived_rank_percentile = perceived_rank_percentile ?? null;
      body.retyped_rank_percentile = retyped_rank_percentile ?? null;
    }

    const response = await customFetch(url, {
      method: "PATCH",
      headers: {
        "Content-Type": "application/json",
      },
      expectedStatusCode: StatusCodes.ACCEPTED,
      serviceName: "SkillsRankingService",
      serviceFunction: "updateSkillsRankingState",
      failureMessage: `Failed to update skills ranking state for session ${sessionId}`,
      body: JSON.stringify(body),
      expectedContentType: "application/json",
    });

    const data = await response.json();

    if (data.experiment_group && !isValidExperimentGroupKey(data.experiment_group)) {
      throw new SkillsRankingError(`Unknown experiment_group '${data.experiment_group}' from API`);
    }

    return {
      session_id: data.session_id,
      experiment_group:
        SkillsRankingExperimentGroups[data.experiment_group as keyof typeof SkillsRankingExperimentGroups],
      phases: data.phase,
      score: data.score,
      cancelled_after: data.cancelled_after ?? null,
      succeeded_after: data.succeeded_after ?? null,
      puzzles_solved: data.puzzles_solved ?? null,
      correct_rotations: data.correct_rotations ?? null,
      clicks_count: data.clicks_count ?? null,
      perceived_rank_percentile: data.perceived_rank_percentile ?? null,
      retyped_rank_percentile: data.retyped_rank_percentile ?? null,
      started_at: data.started_at,
      completed_at: data.completed_at ?? null,
    };
  }

  /**
   * Update only the metrics for a given session without changing the phase.
   * This is used for periodic progress updates.
   *
   * @param sessionId - The ID of the session to update the metrics for.
   * @param metrics - The metrics to update.
   * @returns {Promise<SkillsRankingState>} The updated ranking state for the session.
   */
  async updateSkillsRankingMetrics(sessionId: number, metrics: SkillsRankingMetrics): Promise<SkillsRankingState> {
    const url = `${this.skillsRankingEndpointUrl}/${sessionId}/skills-ranking/state`;

    const response = await customFetch(url, {
      method: "PATCH",
      headers: {
        "Content-Type": "application/json",
      },
      expectedStatusCode: StatusCodes.ACCEPTED,
      serviceName: "SkillsRankingService",
      serviceFunction: "updateSkillsRankingMetrics",
      failureMessage: `Failed to update skills ranking metrics for session ${sessionId}`,
      body: JSON.stringify(metrics),
      expectedContentType: "application/json",
    });

    const data = await response.json();

    if (data.experiment_group && !isValidExperimentGroupKey(data.experiment_group)) {
      throw new SkillsRankingError(`Unknown experiment_group '${data.experiment_group}' from API`);
    }

    return {
      session_id: data.session_id,
      experiment_group:
        SkillsRankingExperimentGroups[data.experiment_group as keyof typeof SkillsRankingExperimentGroups],
      phases: data.phase,
      score: data.score,
      cancelled_after: data.cancelled_after,
      succeeded_after: data.succeeded_after,
      puzzles_solved: data.puzzles_solved,
      correct_rotations: data.correct_rotations,
      clicks_count: data.clicks_count,
      perceived_rank_percentile: data.perceived_rank_percentile,
      retyped_rank_percentile: data.retyped_rank_percentile,
      started_at: data.started_at,
      completed_at: data.completed_at,
    };
  }

  /**
   * Create a throttled function for updating metrics periodically.
   * This sends updates every interval while the user is working, rather than waiting for them to stop.
   *
   * @param sessionId - The ID of the session to update metrics for.
   * @param delay - The throttle delay in milliseconds (default: 2000ms).
   * @returns A throttled function that can be called with metrics to update.
   */
  createThrottledMetricsUpdater(sessionId: number, delay: number = 2000) {
    return createThrottledMetricsUpdater(sessionId, this.updateSkillsRankingMetrics.bind(this), 5000, delay);
  }

  public getConfig(): SkillsRankingConfig {
    return super.getConfig() as SkillsRankingConfig;
  }

  validateConfig(config: any): void {
    super.validateConfig(config);
    // check internal config properties inside the config.config
    if (!config.config.hasOwnProperty("airtimeBudget") || typeof config.config.airtimeBudget !== "string") {
      throw new SkillsRankingError(
        `Invalid configuration for feature ${this.featureId}: airtimeBudget must be a string`
      );
    }
    if (!config.config.hasOwnProperty("jobPlatformUrl") || typeof config.config.jobPlatformUrl !== "string") {
      throw new SkillsRankingError(
        `Invalid configuration for feature ${this.featureId}: jobPlatformUrl must be a string`
      );
    }
  }
}
