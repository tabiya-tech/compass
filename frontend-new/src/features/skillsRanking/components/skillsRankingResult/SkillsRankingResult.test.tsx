// mute the console
import "src/_test_utilities/consoleMock";

import { render, screen } from "src/_test_utilities/test-utils";
import SkillsRankingResult, {
  DATA_TEST_ID,
} from "src/features/skillsRanking/components/skillsRankingResult/SkillsRankingResult";
import { CompareAgainstGroup, ButtonOrderGroup } from "src/features/skillsRanking/types";
import { SkillsRankingService } from "src/features/skillsRanking/skillsRankingService/skillsRankingService";
import { DATA_TEST_ID as CHAT_BUBBLE_DATA_TEST_ID } from "src/chat/chatMessage/components/chatBubble/ChatBubble";
import UserPreferencesStateService from "src/userPreferences/UserPreferencesStateService";
import { SkillsRankingPhase, SkillsRankingState } from "src/features/skillsRanking/types";
import { resetAllMethodMocks } from "src/_test_utilities/resetAllMethodMocks";


describe("SkillsRankingResult tests", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.spyOn(UserPreferencesStateService.getInstance(), "getActiveSessionId").mockReturnValueOnce(1234);
  });

  describe("render tests", () => {
    beforeEach(() => {
      resetAllMethodMocks(SkillsRankingService.getInstance())
      jest.clearAllMocks()
    })
    test("should render component successfully", async () => {
      // GIVEN a skills ranking state
      const givenSkillsRankingState: SkillsRankingState = {
        phase: SkillsRankingPhase.EVALUATED,
        experiment_groups: {
          compare_against: CompareAgainstGroup.AGAINST_OTHER_JOB_SEEKERS,
          button_order: ButtonOrderGroup.SKIP_BUTTON_FIRST,
          delayed_results: false,
        },
        session_id: 1234,
        ranking: "foobar",
        self_ranking: null,
      };

      // WHEN the component is rendered
      render(
        <SkillsRankingResult
          message="Here's what we found"
          skillsRankingState={givenSkillsRankingState}
          isLoading={false}
        />
      );

      // THEN expect the container to be in the document
      const container = screen.getByTestId(DATA_TEST_ID.SKILLS_RANKING_RESULT_CONTAINER);
      expect(container).toBeInTheDocument();
      // AND expect the message bubble to be visible
      expect(screen.getByTestId(CHAT_BUBBLE_DATA_TEST_ID.CHAT_MESSAGE_BUBBLE_CONTAINER)).toBeInTheDocument();
      // AND expect the result text to be in the document
      expect(screen.getByTestId(DATA_TEST_ID.SKILLS_RANKING_RESULT_TEXT)).toBeInTheDocument();
      // AND to match the snapshot
      expect(container).toMatchSnapshot();
      // AND no errors or warnings to have occurred
      expect(console.error).not.toHaveBeenCalled();
      expect(console.warn).not.toHaveBeenCalled();
    });

    test.each([
      [CompareAgainstGroup.AGAINST_OTHER_JOB_SEEKERS, "foobar"],
      [CompareAgainstGroup.AGAINST_JOB_MARKET, "foobar"],
    ])("should render correct result text for %s", async (experimentGroup, rank) => {
      // GIVEN a skills ranking state
      const givenSkillsRankingState: SkillsRankingState = {
        phase: SkillsRankingPhase.EVALUATED,
        experiment_groups: {
          compare_against: experimentGroup,
          button_order: ButtonOrderGroup.SKIP_BUTTON_FIRST,
          delayed_results: false,
        },
        session_id: 1234,
        ranking: rank,
        self_ranking: null,
      };

      // WHEN the component is rendered
      render(
        <SkillsRankingResult
          message="Here's what we found"
          skillsRankingState={givenSkillsRankingState}
          isLoading={false}
        />
      );

      // THEN expect the result text to be in the document
      const resultText = screen.getByTestId(DATA_TEST_ID.SKILLS_RANKING_RESULT_TEXT);
      expect(resultText).toBeInTheDocument();
      // AND expect the result text to contain the correct rank
      expect(resultText).toHaveTextContent(rank);
    });

    test("should show loading state", () => {
      // GIVEN a skills ranking state
      const givenSkillsRankingState: SkillsRankingState = {
        phase: SkillsRankingPhase.EVALUATED,
        experiment_groups: {
          compare_against: CompareAgainstGroup.AGAINST_OTHER_JOB_SEEKERS,
          button_order: ButtonOrderGroup.SKIP_BUTTON_FIRST,
          delayed_results: false,
        },
        session_id: 1234,
        ranking: "foobar",
        self_ranking: null,
      };

      // WHEN the component is rendered with isLoading true
      render(
        <SkillsRankingResult
          message="Here's what we found"
          skillsRankingState={givenSkillsRankingState}
          isLoading={true}
        />
      );

      // THEN expect the loading text and indicator to be in the document
      expect(screen.getByTestId(DATA_TEST_ID.SKILLS_RANKING_LOADING_TEXT)).toBeInTheDocument();
      expect(screen.getByTestId(DATA_TEST_ID.SKILLS_RANKING_LOADING_INDICATOR)).toBeInTheDocument();
      // AND expect the result text to not be in the document
      expect(screen.queryByTestId(DATA_TEST_ID.SKILLS_RANKING_RESULT_TEXT)).not.toBeInTheDocument();
    });
  });
  test.todo("should show delayed result message if the user is assigned delayed result experiment group")
});
