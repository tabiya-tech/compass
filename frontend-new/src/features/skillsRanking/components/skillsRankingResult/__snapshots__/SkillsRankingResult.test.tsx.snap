// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkillsRankingResult tests render tests should render component successfully 1`] = `
<div
  class="MuiBox-root css-1c8zvho"
  data-testid="skills-ranking-result-container-3fb55364-3b63-48db-b1ef-818a4daa5862"
  origin="COMPASS"
>
  <div
    class="MuiBox-root css-u9ttk8"
    data-testid="chat-message-bubble-container-6e685eeb-2b54-432a-8b66-8a81633b3981"
    origin="COMPASS"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-13zm2uq-MuiTypography-root"
      data-testid="chat-message-bubble-message-text-6e685eeb-2b54-432a-8b66-8a81633b3981"
    >
      Here's what we found
    </p>
    <div
      class="MuiBox-root css-0"
      data-testid="chat-message-bubble-message-footer-container-6e685eeb-2b54-432a-8b66-8a81633b3981"
    >
      <div
        class="MuiBox-root css-0"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1czz4sp-MuiTypography-root"
          data-testid="skills-ranking-result-text-3fb55364-3b63-48db-b1ef-818a4daa5862"
        >
          Based on how important your skills are versus other job seekers, you currently rank in the 
          <b>
            foobar
          </b>
           percentile. Do note that this can always change as you continue learning!
        </p>
      </div>
    </div>
  </div>
</div>
`;
